src/script/cpatch.py

```diff
@@ -0,0 +1,727 @@
+import logging
+import logging.config
+import os
+import pathlib
+import shutil
+import subprocess
+import sys
+import tarfile
+import tempfile
+import time
+
+log = logging.getLogger()
+
+
+BASE_IMAGE = "quay.ceph.io/ceph-ci/ceph:main"
+DEST_IMAGE = "ceph/ceph:wip"
+
+# Set CLEANUP to false to skip cleaning up temporary files for debugging.
+CLEANUP = True
+CATCH_ERRORS = True
+
+
+class Component(enum.Enum):
+    """Component to be copied into the new container image."""
+
+    PY_MGR = ("mgr", "python", False)
+    PY_COMMON = ("common", "python", False)
+    CEPHADM = ("cephadm", "python", False)
+    PYBIND = ("bindings", "python", True)
+    DASHBOARD = ("dashboard", "dashboard", True)
+    CORE = ("core", "core", True)
+    CEPHFS = ("cephfs", "core", True)
+    RBD = ("rbd", "core", True)
+    RGW = ("rgw", "rgw", True)
+
+    def __init__(self, name, group, needs_compilation=True):
+        self.component_name = name
+        self.component_group = group
+        self.needs_compilation = needs_compilation
+
+    def __str__(self):
+        return f"{self.component_group}/{self.component_name}"
+
+
+def _exclude_dirs(dirs, excludes):
+    """Exclude directories from a fs walk."""
+    for exclude in excludes:
+        try:
+            del dirs[dirs.index(exclude)]
+        except ValueError:
+            pass
+
+
+def _run(cmd, *args, **kwargs):
+    """Wrapper for subprocess.run with additional logging"""
+    log.debug("running command: %r", cmd)
+    result = subprocess.run(cmd, *args, **kwargs)
+    log.debug("command exited with returncode=%d", result.returncode)
+    return result
+
+
+def check_build_dir(cli_ctx):
+    """Raise an error if the build dir is not populated by a build."""
+    log.debug("checking if build dir is OK")
+    bdir = cli_ctx.build_dir
+    makefile = bdir / "Makefile"
+    ninja = bdir / "build.ninja"
+    if not (makefile.is_file() or ninja.is_file()):
+        log.error(f"neither {makefile} nor {ninja} found")
+        raise ValueError("invalid build dir")
+    return
+
+
+def pull_base_image(cli_ctx):
+    """Pull the base image."""
+    cmd = [cli_ctx.engine, "pull", cli_ctx.base_image]
+    if cli_ctx.root_build:
+        cmd.insert(0, "sudo")
+    _run(cmd).check_returncode()
+
+
+def push_image(cli_ctx):
+    """Push the target image. Assumes login already performed."""
+    cmd = [cli_ctx.engine, "push", cli_ctx.target]
+    if cli_ctx.root_build:
+        cmd.insert(0, "sudo")
+    _run(cmd).check_returncode()
+
+
+class ChangeDir:
+    """Context manager for temporarily changing directory."""
+
+    def __init__(self, path):
+        self.path = path
+        self.orig = None
+
+    def __enter__(self):
+        self.orig = pathlib.Path.cwd().absolute()
+        os.chdir(self.path)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        os.chdir(self.orig)
+
+
+class AddComponentsAction(argparse.Action):
+    """Helper for adding a component to the CLI."""
```

Comment:
What do you mean by "adding a component to the CLI"? Does this let me build a full new executable or something as part of running this?

Reply:
A component is something like *just* cephadm (the binary) or python mgr modules. This class plugs into argparse so we can have CLI options that only choose one component like `--cephadm` but we also want it to work like the previous script which has options like `--py` for all python based components.
This just tries to make it eaiser to build a mix of these single and multiple component CLI options.